### ISSUE TITLE
ASC-4 SESSION MANAGEMENT 

### ISSUE THREAD

**Issue by MenaMachado** [CDH-Studio/I-Talent#354]
**ID:** ASC-4 SESSION MANAGEMENT 

**Description**

If the application has no need to track user behavior prior to authentication, it is safer to defer generating and issuing session IDs until a user has successfully completed the authentication procedure. 

One of the core components of any application is the mechanism by which it controls and maintains the state for a user interacting with it. This is referred to as Session Management and is defined as the set of all controls governing state-full interaction between a user and the application. This control ensures that a verified application (1) creates sessions that are unique to each individual and cannot be guessed or shared and (2) invalidates sessions when they are no longer required and timed out during periods of inactivity. 

Session management go hand in hand with authorization. 

Session Management Best Practices 
1. Enforcing absolute session timeouts 
** If you’re storing session state at the server, you can alleviate the inconvenience of forcing users to start from scratch with their tasks by restoring their session state when they re-authenticate and receive a new session ID. 
** Destroy the expired session ID, but keep the session state around for later use. 
** Protected B applications should limit session IDs to two hours, while for Protected A (or lower) applications, a four hour limit is sufficient. Adjust these values based on the level of risk you’re willing to accept. 
** While you should specify an expiration timestamp on the session ID cookie itself, you must also enforce session timeouts on the server. 
2. Enforcing idle session timeouts 
** Best practice for Protected B applications is an idle timeout of no more than 15 minutes, and a 30-minute timeout for Protected A (or lower) application. 
3. Limiting session concurrency 
** The reasonable limit for your application will depend on the nature of the application. 
4. Mandating secure cookies 
** One of the attributes you can set on the cookie that holds your application’s encrypted session ID is the “Secure” flag, which mandate that the web client uses encrypted communications (SSL and HTTPS) when sending cookie back to the serer along with each request. 
5. Using the HttpOnly flag 
** The HttpOnly flag on a cookie instructs the client that the cookie is only to be used in communications over HTTP protocol. This effectively denies access to the cookie to any client-side code such as JavaScript. 
6. Using cryptographically random session IDs 
** Session ID should be of at least 128 bits in length and generated using secure pseudorandom number generator. 
** Use a cryptographically secure ransom number generator (CSPRNG). Random number sequences can be predictable, and not all random number generators are equal. 
7. Destroying invalidated session IDs 
** Invalidate the session ID on the server. 
** Immediately destroy any related client-side cookie value. Do this by including a Set-Cookie directives in the response with an expiration time set in the past. 
** Delete cookies when the browser closes. 
8. Using encrypted cookies 
** Encrypt any data you store in cookies on the web client. 
** Encrypt any data that is sent with a key that is not accessible to the client, so the client cannot decrypt it. 
** If session state information ever does have to leave the server-side, encrypt it. 
** Use a Hash-based Message Authentication Code (HMAC) to verify the integrity of state information received from the client. 
9. Logging out 
** Provide an explicit logout facility in your application’s UI 
10. Regenerating session IDs on authentication / privilege level change 
** When a user authenticates to a higher level (or even switches roles to access a different area), then the application should grant a new session ID with different rights or update the rights associated with the existing session ID. Updating the rights associated with a session ID is fraught with complexity and potential for introducing vulnerabilities into the application. Generally, it is safer to invalidate an older session ID and issue a new one any time a user changes authentication levels – upward or downward. 
11. Keep all session state information on the server-side 
** Keep as much as you can on the server. Ideally, keep all session state information at the server end. 
** Transmit as little as possible to the client. Ideally, don’t transmit anything other than opaque session ID. 
** Encrypt anything that is sent to the web client for storage and that needs to come back unmodified, using an encryption key that is not available to the client, because the client should never be able to decrypt the encrypted data. 
** Rely non-malleable cryptographic algorithms as a means to detect tampering; let decryption failures trigger immediate invalidation of the session. 
12. Make sure that your web application contains code that validates the session ID every time that protected data or functionality is being requested. 

Sources / References: 
1. Web Application Security, A Beginner's Guide by Bryan Sullivan and Vincent Liu 

**Evidence Required**
If the application uses an existing session management framework, (e.g., IDM), please name the framework and do not provide evidence item #1. 

Please provide: 

1. Documentation explaining the session management policy 

The documentation must identify: 
- If the application enforces absolute session timeouts 
** After how long does the timeout occur? 
** Is the session timeout enforced on the server? 
** Is the session state kept on the server side? 
- If the application enforces idle session timeouts 
** After how long does the timeout occur? 
- If the application allow concurrent sessions 
** How many concurrent sessions are allowed 
- The security flags set on the cookie 
** Secure flag 
** HttpOnly flag 
** Domain flag 
** Path flag 
- How are cryptographically session IDs generated 
** Which cryptographically secure pseudo-random number generator (CSPRNG) is used? 
** How long are the session IDs? 
- How are session IDs invalidated 
** Are session IDs invalidated on the server? 
** Are Set-Cookie directives included in the response with an expiration time set in the past? 
** Are cookies deleted when the browser closes? 
- If encrypted cookies are used 
** Is the encryption key accessible to the client side? 
** Is a Hash-based Message Authentication Code (HMAC) used to verify the integrity of state information received from the client? 
- If the application proved an explicit logout facility in the application`s UI 
- If the application regenerates session IDs on authentication / privilege change 
- If all session state information kept on the server side 
- If the application validates the session ID every time protected data or a protected functionality is being requested 
- Where the sessions IDs are stored 

2. Quality Assurance (QA) test results related to session management (e.g., absolute and idle session timeouts, session concurrency, cookie attributes, session ID randomness, session ID destruction, etc)

**Comment by MenaMachado** [comment]
@BenJeau For this requirement, I think we can skip the first step since we’re using IDM which requires sign in after 30min of inactivity although Best practice recommends 15 for protected B.  Did we implement something in addition to IDM for session mgt? We need to complete item 2 - QA test results. Looking at the examples, are we able to provide test results that prove our application (1) creates sessions that are unique to each individual and cannot be guessed or shared and (2) invalidates sessions when they are no longer required and timed out during periods of inactivity.

**Comment by BenJeau** [comment]
We implemented something for session management, but it is closely tied with IDM - we are only storing the sessions on our servers, but its management seems to be from IDM (I'm not 100% sure on this)

For reference, we are using [express-session](https://github.com/expressjs/session) for session management on our server, and we are using [connect-redis](https://github.com/tj/connect-redis) for storing those sessions in a [redis](https://redis.io/) database. According to what we have set in the configuration of `express-session` right now we are using cookies for our API calls, but just in its most basic form (does not have a maxAge, is not secured, ...). The way session management seems to work on our end right now is, if a new user tries to authenticate themselves, it will create a new entry in the `redis` database and it seems like we are keeping that information forever. But in order to know if the session is valid, we are checking that cookie and keycloak takes precedence over all of this.

So for (1), `express-connect` uses the [uid-safe](https://github.com/crypto-utils/uid-safe) library to create its session ids, and from what I've seen in my `redis` database, its keys are  completely random.

For (2), we would probably need to configure the cookie for our API, but right now, keycloak is taking precedence, so it should be fine (but we should probably still set the configuration for those cookies)
